[RDY] Update lua client

tarruda opened the pull request:
Update to lastest lua client version, which replaces the builtin libuv binding by [luv](https://github.com/luvit/luv)

I'm creating this PR to test on all environments using the new binding, but won't merge until the next version which will replace the lua msgpack code by [libmpack](https://github.com/tarruda/libmpack) and a C UI client.

equalsraf commented:
A windows specific issue with the rockspec is that it calls cmake internally, but the the CMake parameters are not forwarded, this can lead to unexpected issues, for example if building for x86_64, the default cmake generator will build for x86.

Looking at that last commitwith luv+lua-client it almost seems like we are building luv twice (once in a recipe with cmake and a second time via luarocks/rockspec) isn't the first CMake recipe able to do it all? i.e. build and place the modules in the correct lua paths.

tarruda commented:
> Looking at that last commitwith luv+lua-client it almost seems like we are building luv twice (once in a recipe with cmake and a second time via luarocks/rockspec) isn't the first CMake recipe able to do it all? i.e. build and place the modules in the correct lua paths.

Actually, the cmake build is used to install a static version of luv(which I will use later). Also note that the PatchLuv step is only required for mingw, luv now includes the alternate rockspec to build luv without cmake. See https://github.com/luvit/luv/pull/224 for details

equalsraf commented:
@tarruda if the patch is only meant for MinGW/AI_* you can remove it, setting _WIN32_WINNT fixes it for me https://github.com/equalsraf/luv/commit/f037406fe7f02818d2fe6959fbd6ca782f357288, starting a PR upstream in a bit.

tarruda committed:
main.c: Free msgpack packer in --api-info handler to make ASAN happy

tarruda commented:
> @tarruda if the patch is only meant for MinGW/AI_* you can remove it, setting _WIN32_WINNT fixes it for me equalsraf/luv@f037406, starting a PR upstream in a bit.

Thanks for fixing luv for mingw. For now however, I will have to stick with https://github.com/luvit/luv/commit/146f1ce4c08c3b67f604c9ee1e124b1cf5c15cf3 , and there are two reasons for that:

- https://github.com/luvit/luv/commit/2fb919842b5f51ee9038681f4b8cc4156192e838 seems to have broken `BuildLuv.cmake` which uses `WITH_SHARED_LIBUV=ON` and `LUA_BUILD_TYPE=System`(probably because luv scripts for finding libraries won't look in neovim "deps" directory).
- libuv 1.9.0 seems to have broken mingw builds.

When I have time to investigate/fix these issues upstream I will update luv to the latest version.

tarruda committed:
Synchronize shada reset helper with other functional tests

tarruda committed:
Allow a lua debugger to be specified on `functionaltest-lua` target

tarruda committed:
Remove indeterminism in oldfiles_spec.lua

If Nvim is in a "Press ENTER..." screen before the `get_vvar()` call, the test
will hang.

tarruda committed:
Mark cursor_spec.lua indeterministic test as pending

tarruda committed:
Add luv as a third party dependency

Luv is a simple lua binding to libuv, which is now used by neovim lua client.
The bundled luv installation a bit different from other dependencies in that it
is installed two times:

- The "BuildLuv.cmake" script downloads and installs a static version of luv
  using its normal cmake build script. This static version will be used later.
- Luv default rockspec is replaced with the alternate under the "rockspecs"
  directory(the alternate rockspec plays nicer with neovim build system)
- The alternate rockspec is used to build/install the lua module and make it
  available to lua scripts.

tarruda committed:
third-party: Add missing *_host dependencies to luarocks packages

tarruda commented on test/functional/plugin/shada_spec.lua:
```diff
@@ -184,7 +154,7 @@ describe('In autoload/shada.vim', function()
         '  + b                 2',
         '  + c    column       3',
         '  + d                 4',
-      }, {{type=1, timestamp=0, data=st({a=1, b=2, c=3, d=4})}})
```
@ZyX-I can you review this? AFAICT this is not needed for the test(if the purpose is to make it iterate in predictable order), and `__pairs` shouldn't even work in lua 5.1

ZyX-I replied:
AFAIR it is needed for the test because to check that resulting order does not depend on initial order one needs to have known initial order. Using different order will not make test fail if everything is fine, but it may make test *not* fail if something is *not* fine with the ordering. Though it looks like here test was not actually doing what intended: before data reaches tested VimL function it will be transformed into a VimL dictionary: without keeping order obviously.

tarruda commented:
Should be ready, the OSX failures are caused by outdated dependencies.

tarruda committed:
Update lua client to 0.0.1-24

The new version of the lua client uses libmpack as a backend, and some test
scripts had to be updated to reflect that.

lynndylanhurley commented:
After this commit I get the following install error:
https://gist.github.com/lynndylanhurley/da771b8ee594e0ab224da827d01a47b5

OSX 10.11.4
Homebrew 0.9.9 (git revision 36b2a; last commit 2016-04-13)
Homebrew/homebrew-core (git revision 4ce3; last commit 2016-04-13)

justinmk commented:
The homebrew repo needs to be updated.

lynndylanhurley commented:
@justinmk - should I post the issue there instead?

justinmk commented:
@lynndylanhurley Sure
